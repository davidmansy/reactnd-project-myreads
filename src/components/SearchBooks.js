import React, { Component } from 'react';
import SearchBooksBar from './SearchBooksBar';
import * as BooksAPI from '../BooksAPI';
import { debounce } from 'throttle-debounce';
import SearchBooksResults from './SearchBooksResults';

class SearchBooks extends Component {
  state = {
    query: '',
    books: [],
    isLoading: false,
    error: null,
    shelfHasBeenUpdated: false
  };

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onHandleInputChange = query => {
    this.setState({ query, books: [], isLoading: false, error: null }, () => {
      this.search(query);
    });
  };

  search = debounce(400, query => {
    //Call API only if query string is not empty
    if (query.length) {
      this.setState({ isLoading: true }, () => {
        BooksAPI.search(query).then(books => {
          if (this._isMounted) {
            //Condition based on how the API handles error i.e. no error status received
            if (!books.error) {
              this.handleSearchSuccess(books);
            } else {
              this.handleSearchFailure(books.error);
            }
          }
        });
      });
    }
  });

  handleSearchSuccess = books => {
    this.setState(() => ({
      books: this.addBooksCategory(books),
      isLoading: false,
      error: null
    }));
  };

  handleSearchFailure = errorMessage => {
    this.setState({ error: errorMessage, isLoading: false });
  };

  addBooksCategory = books => {
    const { bookCatMap } = this.props;
    return books.map(book => {
      if (bookCatMap[book.id]) {
        book.shelf = bookCatMap[book.id];
      }
      return book;
    });
  };

  handleShelfChange = (book, shelf) => {
    BooksAPI.update(book, shelf).then(() => {
      this.setState(currentState => {
        currentState.books.find(b => b.id === book.id).shelf = shelf;
        currentState.shelfHasBeenUpdated = true;
        return currentState;
      });
    });
  };

  onHandleClose = () => {
    if (this.state.shelfHasBeenUpdated) {
      this.props.onHandleClose();
    }
  };

  render() {
    const { query, books, error, isLoading } = this.state;

    return (
      <div className="search-books">
        <SearchBooksBar
          query={query}
          onHandleChange={event => {
            this.onHandleInputChange(event.target.value);
          }}
          onHandleClose={this.onHandleClose}
        />
        <SearchBooksResults
          error={error}
          isLoading={isLoading}
          books={books}
          handleShelfChange={this.handleShelfChange}
        />
      </div>
    );
  }
}

export default SearchBooks;
