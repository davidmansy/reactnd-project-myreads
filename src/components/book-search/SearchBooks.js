import React, { Component } from 'react';
import SearchBooksBar from './SearchBooksBar';
import * as BooksAPI from '../../data/BooksAPI';
import { debounce } from 'throttle-debounce';
import SearchBooksResults from './SearchBooksResults';

class SearchBooks extends Component {
  state = {
    query: '',
    books: [],
    isLoading: false,
    error: null
  };

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onHandleInputChange = query => {
    this.setState({ query, books: [], isLoading: false, error: null }, () => {
      this.search(query);
    });
  };

  search = debounce(400, query => {
    //Call API only if query string is not empty
    if (query.length) {
      this.setState({ isLoading: true }, () => {
        BooksAPI.search(query).then(books => {
          if (this._isMounted) {
            //Condition based on how the API handles error i.e. no error status received
            if (!books.error) {
              this.handleSearchSuccess(books);
            } else {
              this.handleSearchFailure(books.error);
            }
          }
        });
      });
    }
  });

  handleSearchSuccess = books => {
    this.setState(() => ({
      books: this.assignCategory(books),
      isLoading: false,
      error: null
    }));
  };

  handleSearchFailure = errorMessage => {
    this.setState({ error: errorMessage, isLoading: false });
  };

  assignCategory = books => {
    return books.map(book => {
      const bookWithCat = this.props.booksWithCat.find(b => b.id === book.id);
      book.shelf = bookWithCat ? bookWithCat.shelf : 'none';
      return book;
    });
  };

  handleShelfChange = (book, shelf) => {
    this.setState(currentState => {
      currentState.books.find(b => b.id === book.id).shelf = shelf;
      return currentState;
    });
    this.props.handleShelfChange(book, shelf);
  };

  render() {
    const { query, books, error, isLoading } = this.state;

    return (
      <div className="search-books">
        <SearchBooksBar
          query={query}
          onHandleChange={event => {
            this.onHandleInputChange(event.target.value);
          }}
        />
        <SearchBooksResults
          error={error}
          isLoading={isLoading}
          books={books}
          handleShelfChange={this.handleShelfChange}
        />
      </div>
    );
  }
}

export default SearchBooks;
