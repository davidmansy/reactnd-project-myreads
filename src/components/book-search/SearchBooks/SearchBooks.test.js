import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import SearchBooks from './SearchBooks';
import { MemoryRouter } from 'react-router-dom';
import { cleanup } from 'react-testing-library';

afterEach(cleanup);

describe('SearchBooks', () => {
  const props = {
    rootPath: '/',
    handleShelfChange: () => {},
    booksWithCat: []
  };

  it('renders', () => {
    const container = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <SearchBooks {...props} />
      </MemoryRouter>,
      container
    );
  });

  it('snapshots', () => {
    const testRenderer = TestRenderer.create(
      <MemoryRouter>
        <SearchBooks {...props} />
      </MemoryRouter>
    );
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  //TODO: Test it applies the right category eg assign category function
  //Mock the search to get a list of books
  //Pass a prop as array of books with categories
  //check that the list of books is updated with the categories

  //TODO: Test that if input updated and query string is empty, don't call search
  //TODO: Test that if input updated and query string is not empty, call search with query string

  //Test that if handleSearchSuccess, the state is updated with the returned books
});
