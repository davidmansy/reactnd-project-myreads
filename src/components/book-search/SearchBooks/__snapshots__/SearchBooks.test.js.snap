// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBooks snapshots 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
    data-testid="books-results"
  />
</div>
`;
